diff --git a/src/openslide-vendor-mirax.c b/src/openslide-vendor-mirax.c
index 5cf2b56..6d30816 100644
--- a/src/openslide-vendor-mirax.c
+++ b/src/openslide-vendor-mirax.c
@@ -399,6 +399,17 @@ static bool mirax_detect(const char *filename, struct _openslide_tifflike *tl,
     g_strndup(filename, strlen(filename) - strlen(MRXS_EXT));
   g_autofree char *slidedat_path =
     g_build_filename(dirname, SLIDEDAT_INI, NULL);
+  if (!_openslide_fexists(slidedat_path, &tmp_err)) {
+    g_autofree char *basename = g_path_get_dirname(filename);
+    if (basename) {
+      if (g_str_has_suffix(basename, MRXS_EXT)) {
+        g_free(slidedat_path);
+        g_free(dirname);
+        dirname = g_strndup(basename, strlen(basename) - strlen(MRXS_EXT));
+        slidedat_path = g_build_filename(dirname, SLIDEDAT_INI, SLIDEDAT_INI, NULL);
+      }
+    }
+  }
   if (!_openslide_fexists(slidedat_path, &tmp_err)) {
     if (tmp_err != NULL) {
       g_propagate_prefixed_error(err, tmp_err, "Testing whether %s exists: ",
@@ -1377,10 +1388,22 @@ static bool mirax_open(openslide_t *osr, const char *filename,
   // get directory from filename
   g_autofree char *dirname =
     g_strndup(filename, strlen(filename) - strlen(MRXS_EXT));
+  bool girder_paths = false;
+  GError *tmp_err = NULL;
+  g_autofree char *slidedat_path2 =
+    g_build_filename(dirname, SLIDEDAT_INI, NULL);
+  if (!_openslide_fexists(slidedat_path2, &tmp_err)) {
+    g_free(dirname);
+    g_autofree char *basename = g_path_get_dirname(filename);
+    dirname = g_strndup(basename, strlen(basename) - strlen(MRXS_EXT));
+    girder_paths = true;
+  }

   // first, check slidedat
   g_autofree char *slidedat_path =
-    g_build_filename(dirname, SLIDEDAT_INI, NULL);
+    !girder_paths ?
+    g_build_filename(dirname, SLIDEDAT_INI, NULL) :
+    g_build_filename(dirname, SLIDEDAT_INI, SLIDEDAT_INI, NULL);
   // hash the slidedat
   if (!_openslide_hash_file(quickhash1, slidedat_path, err)) {
     return false;
@@ -1413,7 +1436,7 @@ static bool mirax_open(openslide_t *osr, const char *filename,
   READ_KEY_OR_FAIL(objective_magnification, slidedat, GROUP_GENERAL,
                    KEY_OBJECTIVE_MAGNIFICATION, integer);

-  GError *tmp_err = NULL;
+  tmp_err = NULL;
   int image_divisions =
     g_key_file_get_integer(slidedat, GROUP_GENERAL,
                            KEY_CAMERA_IMAGE_DIVISIONS_PER_SIDE, &tmp_err);
@@ -1503,7 +1526,7 @@ static bool mirax_open(openslide_t *osr, const char *filename,
     g_autofree char *key = g_strdup_printf(KEY_d_FILE, i);
     g_autofree char *name = NULL;
     READ_KEY_OR_FAIL(name, slidedat, GROUP_DATAFILE, key, value);
-    datafile_paths[i] = g_build_filename(dirname, name, NULL);
+    datafile_paths[i] = !girder_paths ? g_build_filename(dirname, name, NULL) : g_build_filename(dirname, name, name, NULL);
   }

   // load data from all slide_zoom_level_section_names sections
@@ -1628,7 +1651,7 @@ static bool mirax_open(openslide_t *osr, const char *filename,
   */

   // read indexfile
-  g_autofree char *index_path = g_build_filename(dirname, index_filename, NULL);
+  g_autofree char *index_path = !girder_paths ? g_build_filename(dirname, index_filename, NULL) : g_build_filename(dirname, index_filename, index_filename, NULL);
   g_autoptr(_openslide_file) indexfile = _openslide_fopen(index_path, err);
   if (!indexfile) {
     return false;

